assembly/src/parsers/u32_ops.rs

```diff
@@ -424,29 +424,54 @@ pub fn parse_u32not(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Ass
 /// 32-bit value.
 ///
 /// VM cycles per mode:
-/// - u32shl: 47 cycles
-/// - u32shl.b: 4 cycles
-/// - u32shl.unsafe: 39 cycles
-pub fn parse_u32shl(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
+/// - u32checked_shl: 47 cycles
+/// - u32checked_shl.b: 4 cycles
+/// - u32unchecked_shl: 40 cycles
+/// - u32unchecked_shl.b: 3 cycles
+/// - u32overflowing_shl: 39 cycles
+/// - u32overflowing_shl.b: 2 cycles
+pub fn parse_u32shl(
+    span_ops: &mut Vec<Operation>,
+    op: &Token,
+    op_mode: U32OpMode,
+) -> Result<(), AssemblyError> {
     let drop_remainder = match op.num_parts() {
         0 => return Err(AssemblyError::invalid_op(op)),
-        1 => {
-            // Assume the dynamic shift value b is on top of the stack.
-            aggregate_power_2(span_ops, false);
-            span_ops.push(Operation::U32assert2);
-            true
-        }
-        2 => match op.parts()[1] {
-            "unsafe" => {
+        1 => match op_mode {
+            U32OpMode::Checked => {
+                // Assume the dynamic shift value b is on top of the stack.
+                aggregate_power_2(span_ops, false);
+                span_ops.push(Operation::U32assert2);
+                true
+            }
+            U32OpMode::Unchecked => {
+                aggregate_power_2(span_ops, true);
+                true
+            }
+            U32OpMode::Overflowing => {
                 aggregate_power_2(span_ops, true);
                 false
             }
-            _ => {
+            _ => return Err(AssemblyError::invalid_op(op)),
+        },
+        2 => match op_mode {
+            U32OpMode::Checked => {
                 let x = parse_u32_param(op, 1, 0, 31)?;
                 span_ops.push(Operation::Push(Felt::new(2u64.pow(x))));
                 span_ops.push(Operation::U32assert2);
                 true
             }
+            U32OpMode::Unchecked => {
+                let x = parse_u32_param(op, 1, 0, u32::MAX)?;
+                span_ops.push(Operation::Push(Felt::new(2u64.pow(x))));
```

Comment:
I'm not sure that in this case using the `push_int_param()` function will be convenient, because we push not just a parsed parameter on the stack, but two to the power of this parameter. It seems that it is possible to adapt `push_int_param()` to this, but is it worth it?

Reply:
Ah - let's keep it as is then.